Wikipedia/Code/SessionsFunnel.swift

```diff
@@ -9,67 +7,59 @@
     private var pageLoadMax: Double?
     private var pageLoadTimes: [Double] = []
     private var pageLoadAverage: Double?
-    
-    private override init() {
-        super.init(schema: "MobileWikiAppiOSSessions", version: 24181799)
-    }
-    
-    private enum Action: String {
+
+    private enum Action: String, Codable {
         case sessionStart = "session_start"
         case sessionEnd = "session_end"
     }
-    
-    private func event(category: EventLoggingCategory, label: EventLoggingLabel?, action: Action, measure: Double? = nil) -> [String: Any] {
-        let category = category.rawValue
-        let action = action.rawValue
-        
-        var event: [String: Any] = ["category": category, "action": action, "primary_language": primaryLanguage(), "is_anon": isAnon]
-        
-        if let label = label?.rawValue {
-            event["label"] = label
-        }
-        if let measure = measure {
-            event["measure_time"] = Int(round(measure))
-        }
-        
-        if let pageLoadMin {
-            event["page_load_latency_min"] = Int(round(pageLoadMin))
-        }
-        
-        if let pageLoadMax {
-            event["page_load_latency_max"] = Int(round(pageLoadMax))
-        }
-        
-        if let pageLoadAverage {
-            event["page_load_latency_average"] = Int(round(pageLoadAverage))
-        }
-        
-        return event
+
+    public struct Event: EventInterface {
+        public static let schema: EventPlatformClient.Schema = .sessions
+        let length_ms: Int?
+        let page_load_latency_ms: Int?
     }
-    
-    override func preprocessData(_ eventData: [AnyHashable: Any]) -> [AnyHashable: Any] {
-        return wholeEvent(with: eventData)
+
+    private func logEvent(measure: Double? = nil) {
+        let measureTime: Int?
+        if let measure {
+            measureTime = Int(round(measure))
+        } else {
+            measureTime = nil
+        }
+
+        let finalEvent = SessionsFunnel.Event(length_ms: measureTime, page_load_latency_ms: Int(round(pageLoadAverage ?? Double())))
+        EventPlatformClient.shared.submit(stream: .sessions, event: finalEvent)
     }
-    
-    @objc public func logSessionStart() {
-        resetSession()
-        log(event(category: .unknown, label: nil, action: .sessionStart))
+
+    @objc public func setSessionStart() {
+        if EventPlatformClient.shared.hasSessionTimedOut {
+            defer {
+                resetSession()
```

Comment:
SessionsFunnel being able to reset EventPlatform's session logic still feels a bit iffy to me and could be the source of future bugs. I'm struggling with how best to clean it up though. Maybe once that is the last problem to solve, we can tackle it in a followup PR.

Reply:
I've separated the sessionID creation logic into another class, but I'm still struggling to find a better way to handle this bit without creating a tight coupling of the new `UserSession` class and the `SessionsFunnel`